subsys/bluetooth/audio/audio_iso.c

```diff
@@ -0,0 +1,143 @@
+/* Common functions for LE Audio services */
+
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "audio_iso.h"
+#include "zephyr/sys/util.h"
+#include "zephyr/sys/util_macro.h"
+
+#define ISO_POOL_SIZE \
+	MIN(CONFIG_BT_ISO_MAX_CHAN, \
+	    (IF_ENABLED(CONFIG_BT_ASCS, \
+			(CONFIG_BT_ASCS_ASE_SNK_COUNT + \
+			 CONFIG_BT_ASCS_ASE_SRC_COUNT)) + \
+	     IF_ENABLED(CONFIG_BT_AUDIO_UNICAST_CLIENT, \
+			(CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_COUNT * \
+			 CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_STREAM_COUNT)) + \
+	     IF_ENABLED(CONFIG_BT_AUDIO_BROADCAST_SOURCE, \
+			(CONFIG_BT_AUDIO_BROADCAST_SRC_COUNT * \
+			 CONFIG_BT_AUDIO_BROADCAST_SRC_STREAM_COUNT)) + \
+	     IF_ENABLED(CONFIG_BT_AUDIO_BROADCAST_SINK, \
+			(CONFIG_BT_AUDIO_BROADCAST_SNK_COUNT * \
+			 CONFIG_BT_AUDIO_BROADCAST_SNK_STREAM_COUNT)) + 0))
+
+static struct bt_audio_iso iso_pool[ISO_POOL_SIZE];
+
+struct bt_audio_iso *bt_audio_iso_new(void)
+{
+	struct bt_audio_iso *audio_iso = NULL;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(iso_pool); i++) {
+		if (atomic_cas(&iso_pool[i].ref, 0, 1)) {
+			audio_iso = &iso_pool[i];
+			break;
+		}
+	}
+
+	if (!audio_iso) {
+		return NULL;
+	}
+
+	(void)memset(audio_iso, 0, offsetof(struct bt_audio_iso, ref));
+
+	return audio_iso;
+}
+
+struct bt_audio_iso *bt_audio_iso_ref(struct bt_audio_iso *audio_iso)
+{
+	atomic_val_t old;
+
+	__ASSERT_NO_MSG(audio_iso);
+
+	/* Reference counter must be checked to avoid incrementing ref from
+	 * zero, then we should return NULL instead.
+	 * Loop on clear-and-set in case someone has modified the reference
+	 * count since the read, and start over again when that happens.
+	 */
+	do {
+		old = atomic_get(&audio_iso->ref);
+
+		if (!old) {
+			return NULL;
+		}
+	} while (!atomic_cas(&audio_iso->ref, old, old + 1));
```

Comment:
> I wonder if it would make sense/possible to merge this part with `bt_conn_ref`.
> 
> Essentially what it is doing is
> 
> ```c
> 	atomic_val_t old;
> 
> 	/* Reference counter must be checked to avoid incrementing ref from
> 	 * zero, then we should return NULL instead.
> 	 * Loop on clear-and-set in case someone has modified the reference
> 	 * count since the read, and start over again when that happens.
> 	 */
> 	do {
> 		old = atomic_get(ref_counter);
> 
> 		if (!old) {
> 			return NULL;
> 		}
> 	} while (!atomic_cas(ref_counter, old, old + 1));
> ```
> 
> The type of the struct that hold the reference counter doesn't really matter, right?

What you mean by `type`?

This code does double-check of reference counter just in case there someone else called the ref/unref (modified the counter basically).

Reply:
By `type` I simply meant that the reference counter handling doesn't care about whether it's a reference counter in a `bt_conn` or a `bt_audio_iso` struct, and that the common code to add a reference if, and only if, the counter is non-zero could be moved to a common reference counter function.

Reply:
But isn't the lifetime of bt_conn different than bt_audio_iso? Having the bt_audio_iso in `listening` state, it has no conn as I understand correctly. Even if it's the same, it would require to provide some kind of `destroy` callback to bt_conn which would require major changes I guess,